# EJERCICIO TITANIC

![](https://upload.wikimedia.org/wikipedia/commons/thumb/f/fd/RMS_Titanic_3.jpg/800px-RMS_Titanic_3.jpg)

Para el ejercicio final, intentá combinar todas las habilidades que aprendiste hasta ahora para cumplir las consignas. **Sentite libre de desviarte de las consignas y explorar los datos que te parezcan interesantes.**

**Link dataset**: https://raw.githubusercontent.com/PPereyraAN/cursos/main/titanic.csv

Este archivo contiene información sobre todos los pasajeros y tripulantes del RMS Titanic. Sus columnas son:

- ``name``: strings with the name of the passenger.
- ``gender``: "male" or "female".
- ``age``: a float with the persons age on the day of the sinking. The age of babies (under 12 months) is given as a fraction of one year.
- ``class``: a string specifying the class (1st, 2nd or 3rd) for passengers or the type of service aboard for crew members.
- ``embarked``: the persons place of of embarkment.
- ``country``: the persons home country.
- ``ticketno``: the persons ticket number (NA for crew members).
- ``fare``: the ticket price (NA for crew members, musicians and employees of the shipyard company).
- ``sibsp``: a number specifying the number if siblings/spouses aboard.
- ``parch``: a number specifying the number of parents/children aboard.
- ``survived``: a string ("no" or "yes") specifying whether the person has survived the sinking.

***Leer el dataset y mostrar:***

- ***Los primeros 5 registros***
- ***Los ultimos 5 registros***

In [6]:
#Importamos nuestro dataset
#Importamos librearia pandas y le asignamos pd como alias
import pandas as pd
url = "https://raw.githubusercontent.com/PPereyraAN/cursos/main/titanic.csv"
titanic = pd.read_csv(url,encoding = 'latin-1')
titanic

name  gender   age             class embarked  \
0                Abbing, Mr. Anthony    male  42.0               3rd        S   
1          Abbott, Mr. Eugene Joseph    male  13.0               3rd        S   
2        Abbott, Mr. Rossmore Edward    male  16.0               3rd        S   
3     Abbott, Mrs. Rhoda Mary 'Rosa'  female  39.0               3rd        S   
4        Abelseth, Miss. Karen Marie  female  16.0               3rd        S   
...                              ...     ...   ...               ...      ...   
2202                Wynn, Mr. Walter    male  41.0         deck crew        B   
2203             Yearsley, Mr. Harry    male  40.0  victualling crew        S   
2204        Young, Mr. Francis James    male  32.0  engineering crew        S   
2205             Zanetti, Sig. Minio    male  20.0  restaurant staff        S   
2206              Zarracchi, Sig. L.    male  26.0  restaurant staff        S   

            country  ticketno   fare  sibsp  parch survived  
0     United States    5547.0   7.11    0.0    0.0       no  
1     United States    2673.0  20.05    0.0    2.0       no  
2     United States    2673.0  20.05    1.0    1.0       no  
3           England    2673.0  20.05    1.0    1.0      yes  
4            Norway  348125.0   7.13    0.0    0.0      yes  
...             ...       ...    ...    ...    ...      ...  
2202        England       NaN    NaN    NaN    NaN      yes  
2203        England       NaN    NaN    NaN    NaN      yes  
2204        England       NaN    NaN    NaN    NaN       no  
2205        England       NaN    NaN    NaN    NaN       no  
2206        England       NaN    NaN    NaN    NaN       no  

[2207 rows x 11 columns]

Una vez importado,primero empezaremos con un analisis general de nuestro dataset, utilizaremos el comando info() que nos dara una descripcion de nuestro dataset, lo cual utilizaremos para comprobar si hay valores nan, o datos que necesiten pasar por algun tipo de analisis. Este tipo de limpieza se realiza con el fin de obtener informacion del mismo dataset y poder sacar promedios, valores medios, max y minimos,etc.

In [3]:
titanic.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2207 entries, 0 to 2206
Data columns (total 11 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   name      2207 non-null   object 
 1   gender    2207 non-null   object 
 2   age       2205 non-null   float64
 3   class     2207 non-null   object 
 4   embarked  2207 non-null   object 
 5   country   2126 non-null   object 
 6   ticketno  1316 non-null   float64
 7   fare      1291 non-null   float64
 8   sibsp     1307 non-null   float64
 9   parch     1307 non-null   float64
 10  survived  2207 non-null   object 
dtypes: float64(5), object(6)
memory usage: 189.8+ KB


Chequeamos el promedio de valores nulos de nuestro dataset. Podremos ver que hay algunos valores nulos, pero son aquellas columnas que corresponden a datos de pasajeros, no del personal propio del titanic. Asi que estos nan conviene dejarlos asi, y filtrarlos en algun otro caso.

In [11]:
(titanic.isna().sum()*100/titanic.shape[0]).sort_values(ascending=False)

fare        41.504304
sibsp       40.779338
parch       40.779338
ticketno    40.371545
country      3.670140
age          0.090621
name         0.000000
gender       0.000000
class        0.000000
embarked     0.000000
survived     0.000000
dtype: float64

Por medio de esta funcion eliminaremos aquellas filas que tengan todos nulos. En este caso, ninguna. Pero igualmente es preferible hacer esto siempre que se haga un analisis

In [13]:
titanic.dropna(axis = 0, inplace=True, how='all')

Procedemos a chequear si tenemos elementos duplicados nuestro dataset. Con la funcion sum haremos una sumatoria de valroes duplicados en todo nuestro dataset

In [16]:
titanic.duplicated().sum()

0

***¿Cuantas filas y columnas tiene el dataset?***

In [14]:
titanic.shape

(2207, 11)

Nuestro dataset posee 2207 filas y 11 columnas

***Cambiar el nombre de la columna ```age``` por ```edad```***

In [14]:
titanic.rename(columns={'age': 'edad'}, inplace=True)
titanic

name  gender  edad             class embarked  \
0                Abbing, Mr. Anthony    male  42.0               3rd        S   
1          Abbott, Mr. Eugene Joseph    male  13.0               3rd        S   
2        Abbott, Mr. Rossmore Edward    male  16.0               3rd        S   
3     Abbott, Mrs. Rhoda Mary 'Rosa'  female  39.0               3rd        S   
4        Abelseth, Miss. Karen Marie  female  16.0               3rd        S   
...                              ...     ...   ...               ...      ...   
2202                Wynn, Mr. Walter    male  41.0         deck crew        B   
2203             Yearsley, Mr. Harry    male  40.0  victualling crew        S   
2204        Young, Mr. Francis James    male  32.0  engineering crew        S   
2205             Zanetti, Sig. Minio    male  20.0  restaurant staff        S   
2206              Zarracchi, Sig. L.    male  26.0  restaurant staff        S   

            country  ticketno   fare  sibsp  parch survived  
0     United States    5547.0   7.11    0.0    0.0       no  
1     United States    2673.0  20.05    0.0    2.0       no  
2     United States    2673.0  20.05    1.0    1.0       no  
3           England    2673.0  20.05    1.0    1.0      yes  
4            Norway  348125.0   7.13    0.0    0.0      yes  
...             ...       ...    ...    ...    ...      ...  
2202        England       NaN    NaN    NaN    NaN      yes  
2203        England       NaN    NaN    NaN    NaN      yes  
2204        England       NaN    NaN    NaN    NaN       no  
2205        England       NaN    NaN    NaN    NaN       no  
2206        England       NaN    NaN    NaN    NaN       no  

[2207 rows x 11 columns]

***Aplicar un ```describe``` y sacar por lo menos 3 conclusiones***

In [18]:
#SACAR LAS CONCLUSIONES
titanic.describe()

edad      ticketno         fare        sibsp        parch
count  2205.000000  1.316000e+03  1291.000000  1307.000000  1307.000000
mean     30.436735  2.842157e+05    33.404760     0.499617     0.385616
std      12.159677  6.334726e+05    52.227592     1.042273     0.866092
min       0.166667  2.000000e+00     3.030500     0.000000     0.000000
25%      22.000000  1.426225e+04     7.180600     0.000000     0.000000
50%      29.000000  1.114265e+05    14.090200     0.000000     0.000000
75%      38.000000  3.470770e+05    31.060750     1.000000     0.000000
max      74.000000  3.101317e+06   512.060700     8.000000     9.000000

El describe puede ser una buena herramienta de analisis de datos, ya que nos da un "paneo" general de nuestro dataset, podremos observar, valores minimos, maximos, promedios,entre otros, de todas las columnas de nuestro dataset. Tambien poder, realizar un analisis estadistico sobre cada columna, llegar a tener una media y poder entender la tendencia que tiene los datos, o su dispersion, a traves de la desviacion de los mismos, y comprobar si los mismos estan mas agrupados a la media o viceversa. Finalmente, podremos comprobar si en nuestro dataset hay valores atipicos, por ahi, podremos comprobar mediante el valor max y minimo que hay un valor desproporcionalmente distinto a los demas y puede llegar a ser un tipo de error, que debe ser puesto a analisis de inmediato. Una vez que se haya tomado la desicion adeacuada, con respecto a ese dato, como por ejemplo modificarlo, podremos continuar. En caso contrario,  ese valor, al no ser tomado en cuenta puede llegar a realizar una desviacion grande de nuestros valores, por ejemplo el de promedio.

***¿Cuales son las personas que tienen los tickets con mayores precio?***

In [9]:
titanic.loc[titanic['fare'].sort_values(ascending=False).index].head(5)

name  gender   age class embarked  \
197       Cardeza, Mrs. Charlotte Wardle  female  58.0   1st        C   
714              Lesueur, Mr. Gustave J.    male  35.0   1st        C   
1248             Ward, Miss. Annie Moore  female  37.0   1st        C   
198   Cardeza, Mr. Thomas Drake Martinez    male  36.0   1st        C   
412                   Fortune, Mrs. Mary  female  60.0   1st        S   

            country  ticketno      fare  sibsp  parch survived  
197   United States   17755.0  512.0607    0.0    1.0      yes  
714             NaN   17755.0  512.0607    0.0    0.0      yes  
1248            NaN   17755.0  512.0607    0.0    0.0      yes  
198   United States   17755.0  512.0607    0.0    1.0      yes  
412          Canada   19950.0  263.0000    1.0    4.0      yes

***Contar la cantidad de personas por genero***

In [22]:
male_count = (titanic['gender'] == 'male').sum()
male_count


1718

In [23]:
female_count = (titanic['gender'] == 'female').sum()
female_count

489

***¿De cuantos paises diferentes provenian las personas?***

In [25]:
titanic['country'].value_counts()

country
England                  1125
United States             264
Ireland                   137
Sweden                    105
Lebanon                    71
Finland                    54
Scotland                   36
Canada                     34
Norway                     26
France                     26
Belgium                    22
Northern Ireland           21
Wales                      20
Bulgaria                   19
Switzerland                18
Channel Islands            17
Croatia (Modern)           12
Croatia                    11
Italy                      11
Spain                       9
India                       8
Hungary                     7
Denmark                     7
Argentina                   7
South Africa                6
Germany                     6
Turkey                      6
Australia                   5
Slovenia                    4
Bosnia                      4
Poland                      3
Austria                     3
Netherlands                 2
Gr

***Encontrar la edad promedio de todas las personas a bordo***

In [15]:
titanic['edad'].mean().round()

30.0

***Filtrar solo los hombres, de primera clase, que sean de Estados unidos o Inglaterra. Finalmente ordenarlos por el precio del ticket***

In [34]:
filtro_titanic = titanic.loc[(titanic['gender'] == 'male' ) & (titanic['class'] == '1st') & (titanic['country'] == ('England' or 'United States')) ,:]
filtro_titanic.sort_values(by='fare', ascending=True).head(5)

name gender   age class embarked  country  \
1134    Smith, Mr. Richard William   male  57.0   1st        S  England   
877   Nicholson, Mr. Arthur Ernest   male  60.0   1st        S  England   
628       Julian, Mr. Henry Forbes   male  50.0   1st        S  England   
130       Borebank, Mr. John James   male  41.0   1st        S  England   
1156     Stead, Mr. William Thomas   male  62.0   1st        S  England   

      ticketno   fare  sibsp  parch survived  
1134  113056.0  26.00    0.0    0.0       no  
877      693.0  26.00    0.0    0.0       no  
628   113044.0  26.00    0.0    0.0       no  
130   110489.0  26.11    0.0    0.0       no  
1156  113514.0  26.11    0.0    0.0       no

***Contar cuantas personas murieron o sobrevivieron que no sean ni de Estados unidos ni de Inglaterra***

In [26]:
filtro_titanic = titanic.loc[~titanic['country'].isin(['United States', 'England'])]
filtro_titanic['survived'].value_counts()

survived
no     553
yes    265
Name: count, dtype: int64

***Encontrar la edad promedio de los hombres a bordo***

In [43]:
filtro_hombres = titanic.loc[(titanic['gender'] == 'male' )]
filtro_hombres['edad'].mean().round()

30.83231351981346

***Seleccionar solo las personas que pertenezca a la clase '3rd', luego contar cuantos sobrevivieron y cuantos murieron.***

In [4]:
filtro_titanic_clase = titanic.loc[titanic['class'] == '3rd']
filtro_titanic_clase['survived'].value_counts()

survived
no     528
yes    181
Name: count, dtype: int64

***Tarifa promedio correspondiente a cada tipo de clase***

In [16]:
categorias = ["1st", "2nd", "3rd"]
titanic_clases = titanic.loc[titanic['class'].isin(categorias),:]

titanic_clases.groupby('class')['fare'].mean()


class
1st    89.017344
2nd    21.503081
3rd    12.927864
Name: fare, dtype: float64

***Encuentre la tasa de supervivencia promedio correspondiente a cada tipo de cabina***

In [17]:
titanic['embarked'].value_counts()

embarked
S    1616
C     271
B     197
Q     123
Name: count, dtype: int64

Quiero crear una funcion que me transforme de mi dataset los "yes" en True y los "no" en False. Guardaria una columna nueva y borraria la anterior ya que poner un booleano me parece mucho mas descriptivo que un string en el caso de que haya sobrevivido. Una vez realizado esto procedo a borrar mi  columna survived

In [23]:
def transformar_supervivencia(valor):
    return True if valor == 'yes' else False

In [25]:
cabinas = ["S","C","B","Q"]
titanic_cabinas = titanic.loc[titanic['embarked'].isin(cabinas),:]

titanic_cabinas['supervivencia_bool'] = titanic_cabinas['survived'].apply(transformar_supervivencia)

# Eliminar la columna 'survived'
titanic_cabinas.drop(columns=['survived'], inplace=True)

titanic_cabinas.groupby('embarked')['supervivencia_bool'].mean().round(2)



embarked
B    0.22
C    0.56
Q    0.36
S    0.29
Name: supervivencia_bool, dtype: float64